src/main/EchoServer.java

```diff
@@ -1,26 +1,38 @@
 package main;
 
+import java.io.InputStream;
+import java.io.PrintStream;
 import java.util.Scanner;
 
 public class EchoServer {
 
-    Scanner in = new Scanner(System.in);
+    private PrintStream output;
+    private Scanner input;
 
-    public void promptToType() {
-        System.out.println("Hi! Please type anything:");
+    public EchoServer(PrintStream output, InputStream input) {
+        this.output = output;
+        this.input = new Scanner(input);
     }
 
-    public String getWords() {
-        String words = in.nextLine();
+    public void run() {
+        promptToType();
+        printReversedEcho(getWords());
+    }
+
+    private void promptToType() {
+        this.output.println("Hi! Please type anything:");
+    }
+
+    private String getWords() {
+        String words = input.nextLine();
         while (words.length() == 0) {
-            System.out.println("Please type at least 1 word:");
-            words = in.nextLine();
+            this.output.println("Please type at least 1 word:");
+            words = input.nextLine();
         }
         return words;
     }
 
-    public void printReversedEcho(String usersWords) {
-        System.out.println(new StringBuilder(usersWords).reverse().toString());
+    private void printReversedEcho(String usersWords) {
+        this.output.println(new StringBuilder(usersWords).reverse().toString());
```

Comment:
same as above. If you really want to use `this`, then be consistent and use it on line 30 for `input` as well.